crates/client/rpc/src/lib.rs

```diff
@@ -1058,17 +998,30 @@ where
         })?;
         let chain_id = Felt252Wrapper(self.chain_id()?.0);
 
-        let message = message.try_into().map_err(|e| {
-            error!("Failed to convert MsgFromL1 to UserTransaction: {e}");
-            StarknetRpcApiError::InternalServerError
-        })?;
+        let transaction = {
+            let calldata = std::iter::once(Felt252Wrapper::from(message.from_address).into())
+                .chain(message.payload.into_iter().map(|felt| Felt252Wrapper::from(felt).into()))
+                .collect();
+            let tx = starknet_api::transaction::L1HandlerTransaction {
+                version: TransactionVersion::ZERO,
+                nonce: Nonce(StarkFelt::ZERO),
```

Comment:
hardcoced nonce ?

Reply:
yeah, the nonce is not really used by the blockifier during the execution of l1 tx:

```rust
impl<S: StateReader> ExecutableTransaction<S> for L1HandlerTransaction {
    fn execute_raw(
        self,
        state: &mut TransactionalState<'_, S>,
        block_context: &BlockContext,
        _charge_fee: bool,
        _validate: bool,
    ) -> TransactionExecutionResult<TransactionExecutionInfo> {
        let tx_context = Arc::new(block_context.to_tx_context(&self));

        let mut execution_resources = ExecutionResources::default();
        let mut context = EntryPointExecutionContext::new_invoke(tx_context.clone(), true)?;
        let mut remaining_gas = block_context.versioned_constants.tx_initial_gas();
        let execute_call_info =
            self.run_execute(state, &mut execution_resources, &mut context, &mut remaining_gas)?;
        let l1_handler_payload_size = self.payload_size();

        let (ActualCost { actual_fee, da_gas, actual_resources }, _bouncer_resources) =
            ActualCost::builder_for_l1_handler(tx_context, l1_handler_payload_size)
                .with_execute_call_info(&execute_call_info)
                .try_add_state_changes(state)?
                .build(&execution_resources)?;

        let paid_fee = self.paid_fee_on_l1;
        // For now, assert only that any amount of fee was paid.
        // The error message still indicates the required fee.
        if paid_fee == Fee(0) {
            return Err(TransactionFeeError::InsufficientL1Fee { paid_fee, actual_fee })?;
        }

        Ok(TransactionExecutionInfo {
            validate_call_info: None,
            execute_call_info,
            fee_transfer_call_info: None,
            actual_fee: Fee::default(),
            da_gas,
            actual_resources: actual_resources.clone(),
            revert_error: None,
            bouncer_resources: actual_resources,
        })
    }
}
```

As you can see the `_validate` argument is ignored, nonce is not used and fees are barely checked. It look like the protocol does not contains Nonce verification for l1_handler_tx 🤷‍♂️

So for the the `estimate_message_fee` RPC which contains no nonce, we can just use whatever....